src/tap/cmd.rs

```diff
@@ -98,12 +99,27 @@ async fn run(
         );
     };
 
+    let start_time = Instant::now();
+    let stream_duration = match opts.duration {
+        Some(duration) => Duration::from_millis(duration),
+        None => Duration::MAX,
+    };
+
     // Loop over the returned results, printing out tap events.
     #[allow(clippy::print_stdout)]
     #[allow(clippy::print_stderr)]
     loop {
-        let message = stream.next().await;
-        if let Some(Some(res)) = message {
+        if start_time.elapsed() >= stream_duration {
+            return exitcode::OK;
+        }
+
+        let message = 
+            timeout(stream_duration - start_time.elapsed(), stream.next()).await;
```

Comment:
Both of these calls to `elapsed` end up invoking a system call. While it is very fast, we should probably cache that result, something like:
```suggestion
        let elapsed = start_time.elapsed();
        if elapsed >= stream_duration {
            return exitcode::OK;
        }

        let message = timeout(stream_duration - elapsed, stream.next()).await;
```